File: src/javascript/algorithms/array/25.destinationCity/destinationCity.js
function destinationCity(paths){
    let fromCities = new Set();

    for(const [from, to] of paths) {
        fromCities.add(from);
    }

    for(const [from, to] of paths) {
        if(!fromCities.has(to)) {
            return to;
        }
    }
}

const paths1 = [["Hyderabad","KL"],["KL","Singapore"],["Singapore","Sydney"]];
const paths2 = [["New Jersey","Austin"],["New York","New Jersey"],["Austin","Dallas"]];
const paths3 = [["Dallas", "London"]];

console.log(destinationCity(paths1));
console.log(destinationCity(paths2));
console.log(destinationCity(paths3));

File: src/python/algorithms/array/25.destinationCity/destination_city.py
"""
Destination City

You are given an array of paths where paths[i] = [from, to].
Return the destination city, i.e., the city without any path outgoing to another city.

Time Complexity: O(n)
Space Complexity: O(n)
"""


def destination_city(paths):
    """
    Using set to track cities with outgoing paths.
    TC: O(n), SC: O(n)
    """
    from_cities = set()
    
    for from_city, to_city in paths:
        from_cities.add(from_city)
    
    for from_city, to_city in paths:
        if to_city not in from_cities:
            return to_city
    
    return None


# Test cases
if __name__ == "__main__":
    paths1 = [["Hyderabad", "KL"], ["KL", "Singapore"], ["Singapore", "Sydney"]]
    paths2 = [["New Jersey", "Austin"], ["New York", "New Jersey"], ["Austin", "Dallas"]]
    paths3 = [["Dallas", "London"]]
    
    print(f"Destination: {destination_city(paths1)}")  # Sydney
    print(f"Destination: {destination_city(paths2)}")  # Dallas
    print(f"Destination: {destination_city(paths3)}")  # London
